### ISSUE TITLE
Add `VictoryBoxPlot`

### ISSUE THREAD

**Issue by dtelad11** [FormidableLabs/victory#874]
Currently, `VictoryCandlestick` supports four types of data: open, close, high, and low.

One common type of plot which could have been drawn using `VictoryCandlestick` is a box plot (https://en.wikipedia.org/wiki/Box_plot). However, for a box plot, we require an additional "center" data type. Are there any plans to add such a data type? Could you recommend how to mimic it using another component?

Thank you!

**Comment by boygirl** [comment]
@dtelad11 Interesting idea! I think I would prefer to add an actual `VictoryBoxPlot` component rather than add something to `VictoryCandlestick`. In the meantime, you could absolutely make a custom data component that would render both a candlestick and the additional line you need to make a boxplot. Here's a [guide on custom components](https://formidable.com/open-source/victory/guides/custom-components) to get you started. You can also come ask questions in the [gitter channel](https://gitter.im/FormidableLabs/victory) if you get stuck 

I'm going to change the title of this issue to `VictoryBoxPlot` and flag it as a feature request :)

**Comment by dtelad11** [comment]
Thank you!

**Comment by ersinciftci** [comment]
Looking forward to this feature to convert my d3 box plots to Victory.

**Comment by boygirl** [comment]
Implementation details:

<img width="330" alt="screen shot 2018-01-16 at 10 13 14 am" src="https://user-images.githubusercontent.com/3719995/35004704-f11ce5b4-faa5-11e7-9c8f-a3fc7cab6ba5.png">

This component should take data in the standard form (except with expected repeat x-values)
```js
data={[
 { x: 1, y: 1 }, { x: 1, y: 2 }, { x: 1, y: 3 }, ...
]}
```
Calculate median and quartile values from data: https://en.wikipedia.org/wiki/Quartile

processed into the form.

```js
{ x, min, q1, median, q3, max }
```

This component should take a `horizontal` prop. When true, data will expect repeat y values, and the final processed form will be
```js
{ y, min, q1, median, q3, max }
```

Final processed data should be used for evaluating any props that are a function of data ( _i.e._ styles, labels). Since we're expecting multiple data points for a single box plot element, extra attributes on the data object (_i.e._ label, fill, etc) should be ignored.

rather than the usual `parent`, `data`, `labels`, we will have individual namespaces for each
`min`, `q1`, `median`, `q3`,`max`, `minLabels`, `maxLabels`, `q1Labels`, `q3Labels`,`medianLabels`. 

so `style` will have the shape:
```
style={{
  parent, min, max, q1, q3, median, minLabels, maxLabels, q1Labels, q3Labels, medianLabels
}}
```
and subcomponents will be as follows:
```
containerComponent: <VictoryContainer/>
groupComponent: <g/>
minComponent: <Whisker/> // new primitive component should render a line and crosshair
maxComponent: <Whisker/>
q1Component: <Box/> // new primitive component should render a rect
q3Component: <Box/>
medianComponent: <Line/>
minLabelComponent: <VictoryLabel/>
maxLabelComponent: <VictoryLabel/>
q1LabelComponent: <VictoryLabel/>
q2LabelComponent: <VictoryLabel/>
medianLabelComponent: <VictoryLabel/>
```

Misc requirements:

- user configureable box widths
- configureable crosshair widths (the line at min and max)
- `labelOrientation` prop with options "top", "bottom", "left", "right". Default value will depend on whether the box plots are horizontal

**Comment by jjgao** [comment]
This is an exciting feature and we would definitely use it. Thanks for implementing it?

On our website (cbioportal.org), we conbine boxplot with scatter plot.

![image](https://user-images.githubusercontent.com/840895/35016446-920a6394-fae6-11e7-8747-0e602cac8ddc.png)

https://stackoverflow.com/questions/23675735/how-to-add-boxplots-to-scatterplot-with-jitter

Is it possible to add an option to hide the outliers so that we can overlay a scatter plot on top? It would be even better if you can add an option to support the plot above directly.

**Comment by dtelad11** [comment]
This looks fantastic, thank you!

Just to make sure I understand, the current planned implementation will take the raw data, calculate the statistics, and display the boxplot based on the statistics?

If yes -- would it be possible to support another implementation, where the developer specifies data of the form

```
{ x, min, q1, median, q3, max }
```

In other words, where I could pass the statistics and have the boxplot display appropriately?

The reason I'm asking is for cases where the data set has hundreds of thousands of points. I would prefer to calculate the statistics on the backend rather than pass them on to the frontend.

**Comment by boygirl** [comment]
@dtelad11 yes, that makes a ton of sense. I'll edit the implementation details to reflect that

**Comment by boygirl** [comment]
@jjgao yes, the chart above will be possible with the implementation described above. Outliers will not be handled by `VictoryBoxPlot`, and you will be able to combine it with any other chart type you like.

**Comment by boygirl** [comment]
Released in `victory@0.25.7` Docs to be updated by Monday April 2

**Comment by dtelad11** [comment]
Thank you, this is a fantastic update. Looking forward to toying with the new chart.
